### ISSUE TITLE
[BUG] ggml_cuda_op_mul_mat()  SIGSEGV   on several GPU's

### ISSUE THREAD

**Issue by kotee4ko** [LostRuins/koboldcpp#483]
Hello, thanks for awesome project.

```cpp

ggml_cuda_op_mul_mat() {
 ...
                // copy dst to host or other device if necessary
                if (!dst_on_device) {
                    void * dst_off_device;
                    cudaMemcpyKind kind;
                    if (dst->backend == GGML_BACKEND_CPU) {
                        dst_off_device = dst->data;
                        kind = cudaMemcpyDeviceToHost;
                    } else if (dst->backend == GGML_BACKEND_GPU) {
                        dst_off_device = dst_extra->data_device[g_main_device];
                        kind = cudaMemcpyDeviceToDevice;
                    } else {
                        GGML_ASSERT(false);
                    }
                    if (split) {
                        // src0 = weight matrix is saved as a transposed matrix for better memory layout.
                        // dst is NOT transposed.
                        // The outputs of matrix matrix multiplications can therefore NOT simply be concatenated for >1 GPU.
                        // Instead they need to be copied to the correct slice in ne0 = dst row index.
                        // If dst is a vector with ne0 == 1 then you don't have to do this but it still produces correct results.
                        float * dhf_dst_i = (float *) ((char *) dst_off_device + i02*nb2 + i03*nb3);
                        GGML_ASSERT(dst->nb[1] == ne0*sizeof(float));
                        dhf_dst_i += src1_col_0*ne0 + row_low[id];
                        CUDA_CHECK(cudaMemcpy2DAsync(dhf_dst_i, ne0*sizeof(float), dst_dd_i, row_diff*sizeof(float),
                                                    row_diff*sizeof(float), src1_ncols, kind, stream));
                    } else {
                        float * dhf_dst_i = (float *) ((char *) dst_off_device + i02*nb2 + i03*nb3);
                        GGML_ASSERT(dst->nb[1] == ne0*sizeof(float));
                        dhf_dst_i += src1_col_0*ne0;
                        CUDA_CHECK(cudaMemcpyAsync(dhf_dst_i, dst_dd_i, src1_ncols*ne0*sizeof(float), kind, stream));
                    }
                }

                // add event for the main device to wait on until other device is done
                if (split && (id != g_main_device || is != 0)) {
                    CUDA_CHECK(cudaEventRecord(src0_extra->events[id][is], stream));
                }
            }
   ...
   
 }
 
 ```
 
 if split is true, than call to cudaMemcpy2DAsync() resulting in SIGSEGV. 
 ========================= ROCm System Management Interface =========================
=================================== Concise Info ===================================
GPU  Temp (DieEdge)  AvgPwr  SCLK    MCLK    Fan    Perf  PwrCap  VRAM%  GPU%  
0    45.0c           13.0W   300Mhz  945Mhz  9.41%  auto  110.0W   90%   0%    
1    35.0c           15.0W   300Mhz  945Mhz  10.2%  auto  110.0W   79%   0%    
====================================================================================
=============================== End of ROCm SMI Log ================================

any ideas how to fix? thanks

**Comment by LostRuins** [comment]
Is this a new issue? Are you running from this repo or from https://github.com/YellowRoseCx/koboldcpp-rocm ?

**Comment by kotee4ko** [comment]
The issue isn't new, but yes, I run from latest https://github.com/YellowRoseCx/koboldcpp-rocm.

Speaking true, I even don't know where this issue should be published - amd, ggml, or one of three llama.cpp forks, you know...
